### ISSUE TITLE
Low performance in offline benchmark in the Flutter app

### ISSUE THREAD

**Issue by d-uzlov** [mlcommons/mobile_app_open#159]
https://github.com/mlcommons/mobile_app_open/issues/111#issuecomment-985248507
>But, looks like that the configuration for offline case has some changes, because I am getting very low score, ~170QPS instead of ~2500QPS (v1.1 submission).

**Comment by d-uzlov** [comment]
I compared the performance of the tflite backend between the old app and the Flutter app, and I found that the Flutter is ~1.5 times slower in every benchmark.
I think this is caused by thread priority.
I wasn't able to find any info on thread priority in Flutter, but when I changed the code to create a new native thread for each benchmark, I got the same scores as in Android app. Seems like Flutter sets a low priority for its threads.

I don't have the hardware to test other backends but I think we should see improvements in every backend.

I'll upload a new APK later to test if the performance issues are completely fixed.

**Comment by d-uzlov** [comment]
>I think this is caused by thread priority.

After further research I found out that this is not the case.
Turns out my Note20 Ultra throttles when it heats up to ~40°C. I would never think that a phone that's barely warm could throttle so I thought that my changes caused difference in scores, but after a lot of testing I'm sure that my changes did nothing.

**Comment by d-uzlov** [comment]
@relja128 @Mostelk @mohitmundhragithub @freedomtan 
Could you test this new Android build of the Flutter app?
https://drive.google.com/file/d/1bfnafC5xMZm2Ec1iQv7UAqPYGC46fkh6/view?usp=sharing

**Comment by d-uzlov** [comment]
I added few missing dependencies, which could prevent backends from loading, and [hopefully] fixed config for offline benchmark.

This is a debug build. This should not affect scores in any meaningful way because backends are built in release mode, only UI is in debug mode.
Release builds don't write much logs. This debug build should write all relevant logs.
If something is wrong with performance, please use `adb logcat` to get application logs.
You can search for `checking backend` phrase in logs to check if the app actually loaded and used expected backend, and if something went wrong, there should be an error message with explanation.

**Comment by mohitmundhragithub** [comment]
![image](https://user-images.githubusercontent.com/1549325/147631062-124dbede-5b30-4fcf-b782-7e245af939a1.png)
We tried running the shared apk, but seeing this black screen in the app. The app, which was shared last time worked for us... but not this one.

**Comment by d-uzlov** [comment]
@mohitmundhragithub this error can happen if there is no internet access and the app can't download list of benchmarks.
Could you check if the phone you used has internet access?

**Comment by AasheeshT** [comment]
[flutter_debug_3.txt](https://github.com/mlcommons/mobile_app_open/files/7787616/flutter_debug_3.txt)
Logs with offline mode. We still observe low numbers.

**Comment by d-uzlov** [comment]
@AasheeshT, thank you for logs.
I see that the scores are improved but the app passes incorrect batch size to the backend, so the scores are not as high as they should be.
Here is a new APK with a fix: https://drive.google.com/file/d/1c_dLNxRaQgGuD5sBYs1PDS3z2p45CZ_H/view?usp=sharing

I don't know why the app could crash. It looks like there are some issues with either native code itself or with the Dart code that calls it, but judging by the logs app should have finished working with native code by the time of the crash.
On my phone I haven't encounter any crashes.
Does the app crash every time or only sometimes?
Does the app crash only in offline benchmark or in all benchmarks?

**Comment by AasheeshT** [comment]
@d-uzlov Thanks for the apk. Performance has improved with correct batchsize but app still crashes. The crash is regular and with all the benchmarks.

**Comment by anhappdev** [comment]
I would test the app with different backends to see if the crash occurred only with certain backend.
I don't know if it makes a different, but the QTI backend now uses SNPE SDK version `1.54.2`. The last benchmark submission with Android app used the version `1.48.0`

**Comment by d-uzlov** [comment]
@AasheeshT, could you please test if some of these 2 apks fix crashes?
[mlperfbench-v3.0.6-debug+old_cpp_lib.apk](https://drive.google.com/file/d/1K968amf8cQ-bNE1J78UO1Svp3DmaWSvZ/view?usp=sharing)
[mlperfbench-v3.0.6-release+new_cpp_lib.apk](https://drive.google.com/file/d/1GE67dDA6hPRGDH0WOnnzDCN4iX0ioPsB/view?usp=sharing)

Also, what happens if you run several benchmarks at once? Does the app crash after first benchmark finishes or only after last benchmark?

**Comment by d-uzlov** [comment]
Also, if APKs above don't help, could you upload logs after this APK crashes?
[mlperfbench-v3.0.6-debug+print.apk](https://drive.google.com/file/d/1oTP7x_0BSyncRclhaTYcoAq3Eo9pFJNf/view?usp=sharing)

**Comment by AasheeshT** [comment]
@d-uzlov Crash occurs in both the apks. Below are the logs :
[old_cpp_lib.txt](https://github.com/mlcommons/mobile_app_open/files/7792012/old_cpp_lib.txt)
[new_cpp_lib.txt](https://github.com/mlcommons/mobile_app_open/files/7792013/new_cpp_lib.txt)

**Comment by mohitmundhragithub** [comment]
@d-uzlov, can you please share the PR which fixed the offline case performance issue?

**Comment by d-uzlov** [comment]
@mohitmundhragithub, I haven't made a PR yet, but all the changes I made related to this issue are contained in the [159/perf](https://github.com/mlcommons/mobile_app_open/tree/159/perf) branch.

**Comment by d-uzlov** [comment]
@AasheeshT, could you test if crash occurs in these apks?
[01-+qti-88955ab56a672d6187f697948f6b3dd7f9a43f3d.apk](https://drive.google.com/file/d/19_KlL-3LYgMaOhDkZtSHb3H-p6gkCmUI/view?usp=sharing)
[02-+qti-102af7ac828aecc8a473acc66ddc1ed37f721f59.apk](https://drive.google.com/file/d/17CtYbRXGCDCNY2aFtEwg19ocyqKkFQ-m/view?usp=sharing)
[03-+qti-ef5d75609f27885a52d9b321a92c8b3d8c3b72c0.apk](https://drive.google.com/file/d/1608UIqN1jl_MR2wqPcJQCG_oluM8pnl9/view?usp=sharing)

**Comment by AasheeshT** [comment]
Thanks @d-uzlov for the apks. The crash occurs on all the apks. Following are the logs :
[01.txt](https://github.com/mlcommons/mobile_app_open/files/7852564/01.txt)
[02.txt](https://github.com/mlcommons/mobile_app_open/files/7852565/02.txt)
[03.txt](https://github.com/mlcommons/mobile_app_open/files/7852567/03.txt)

**Comment by AasheeshT** [comment]
Also the above apks don't have the batchsize fix for offline mode. Can you please tell what fix was done earlier to resolve wrong batch size in offline mode.

**Comment by d-uzlov** [comment]
@AasheeshT, all the changes I made related to this issue are contained in the [159/perf](https://github.com/mlcommons/mobile_app_open/tree/159/perf) branch.

**Comment by d-uzlov** [comment]
@AasheeshT, could you test this apk?
https://drive.google.com/file/d/1JRAF6gzO1GPd8SBgcKoPoWsp6Uh_KD0F/view?usp=sharing

**Comment by AasheeshT** [comment]
This also crashes. Here are the logs :
[04.txt](https://github.com/mlcommons/mobile_app_open/files/7853292/04.txt)

**Comment by d-uzlov** [comment]
@mohitmundhragithub, did you encounter any crashes when you tested the app here?
https://github.com/mlcommons/mobile_app_open/issues/111#issuecomment-985248507

**Comment by mohitmundhragithub** [comment]
As i remember, i didnt see any crashes with the other apk.

**Comment by mohitmundhragithub** [comment]
I think there were random crashes with the other apk as well.

**Comment by d-uzlov** [comment]
@AasheeshT, could you test this APK?
https://drive.google.com/file/d/15Pywfu6RM1tZwLvhyQB7ajDfgjKNoFl-/view?usp=sharing
This is the old Android application. I want to check that the issues is specific to the Flutter app or to something from my build environment.

**Comment by d-uzlov** [comment]
Could you also check this new build in the Releases page?
https://github.com/mlcommons/mobile_app_open/releases/tag/v3.0.0-flutter-android-alpha-4

**Comment by freedomtan** [comment]
qualcomm and mtk to test it. Samsung will revisit this after submission. We close this one for now.
